tests/txTests.ts

```diff
@@ -53,21 +62,145 @@ const runTests = async (file: string) => {
+            const fixture = new RevokeFixture(hash);
+            (fixture.oldCip68Datum.constructor_0[2] as any).virtual = { 
+                public_mint: 1, /// public = true
+                expires_slot: Date.now()
+            }
+            const initialized =  await fixture.initialize();
+            initialized.inputs?.splice(1, 1); /// remove 222 root handle from input
+            initialized.outputs = []; /// remove 222 root handle from input
+            return initialized;
+        }, (fixture) => setupRevokeTx(fixture, BigInt(getSlotNumberFromDate(new Date(Date.now() - 1_000_000))))),
+    false, "Publicly minted Virtual SubHandle hasn't expired"),
+
+    // should Deny revoke if NOT signed by root or admin 
+    await tester.test("REVOKE", "private and not signed by root", new Test(program, async (hash) => {
+            const fixture = new RevokeFixture(hash);
+            const initialized =  await fixture.initialize();
+            initialized.inputs?.splice(1, 1); /// remove 222 root handle from input
+            initialized.outputs = []; /// remove 222 root handle from input
+            return initialized;
+        }, (fixture) => setupRevokeTx(fixture)),
+    false, "Publicly minted Virtual SubHandle hasn't expired"),
+
+    // UPDATE - SHOULD APPROVE - should only update if private
+    await tester.test("UPDATE", "private mint", new Test(program, async (hash) => {
+        const fixture = new UpdateFixture(hash);
+        return await fixture.initialize();
+    })),
+
+    // should Deny if we update pz rather than virtual & resolved address
+    await tester.test("UPDATE", "update pz", new Test(program, async (hash) => {
+        const fixture = new UpdateFixture(hash);
+        (fixture.newCip68Datum.constructor_0[2] as any).portal = "ipfs://new_cid"; /// update pz
+        return await fixture.initialize();
+    }), false, "Restricted changes are not allowed"),
+
+    // should Deny if we update nft rather than virtual & resolved address
+    await tester.test("UPDATE", "update nft", new Test(program, async (hash) => {
+        const fixture = new UpdateFixture(hash);
+        (fixture.newCip68Datum.constructor_0[0] as any) = {
+            ...defaultNft,
+            image: "ipfs://new_image"
+        }; /// update nft
+        return await fixture.initialize();
+    }), false, "Restricted changes are not allowed"),
+
+    // should Deny update if assignee NOT signed when public, and root handle not signed
+    await tester.test("UPDATE", "assignee not signed", new Test(program, async (hash) => {
+        const fixture = new UpdateFixture(hash);
+        (fixture.oldCip68Datum.constructor_0[2] as any).virtual.public_mint = 1; /// make public
+        (fixture.newCip68Datum.constructor_0[2] as any).virtual.expires_slot = Date.now(); /// not extend
+        fixture.inputs?.splice(1, 0); /// remove 222 root_handle in inputs
+        fixture.outputs?.splice(0, 0); /// remove 222 root_handle in outputs
```

Comment:
In JavaScript, the second param in splice is the number of elements, yeah? So `0` does nothing?

Reply:
Whoops....
What am I thinking while typing that line? 🤣